src/server/database/PostgreSQL.ts

```
@@ -1,15 +1,23 @@
+import type * as pg from 'pg';
 import {IDatabase} from './IDatabase';
 import {IGame, Score} from '../IGame';
 import {GameOptions} from '../game/GameOptions';
 import {GameId, ParticipantId} from '../../common/Types';
 import {SerializedGame} from '../SerializedGame';
-import {Pool, ClientConfig, QueryResult} from 'pg';
+let Pool: typeof pg.Pool | undefined;
+
+try {
```

Why not let this throw?

---

If you locally run a `rm -rf node_modules/pg` that may show the problem I am working around here. This will remove the postgres dependency leaving you in situation of a user that was unable to install postgres. If you then build and start the server with the logging here you will see the info logged.

```
[2023-09-16 12:41:16 MDT] postgres not installed, use another database or npm install pg
```

If instead this were to throw the server would be unable to start as the exception would be thrown once we required the PostgreSQL module which in turn will throw this exception.

Another option would be to transition towards dynamic imports versus the static imports we use today. At that point we could only call to require PostgreSQL if the user defined the environment variable to use PostgreSQL. Changing to dynamic imports would require us to introduce an await somewhere so we had time to load the module before using it.

---

Perhaps the app should only try to load Postgres when the environment variable is set.  Or when the supplied class is instantiated.

My concern about an ignorable warning  is that it puts users who _want_ postgres in a position where the app will silently fall back on a database they don't want.

---

Perhaps the app should only try to load Postgres when the environment variable is set.  Or when the supplied class is instantiated.

My concern about an ignorable warning  is that it puts users who _want_ postgres in a position where the app will silently fall back on a database they don't want.

---

Perhaps the app should only try to load Postgres when the environment variable is set.  Or when the supplied class is instantiated.

My concern about an ignorable warning  is that it puts users who _want_ postgres in a position where the app will silently fall back on a database they don't want.

---

I will re-factor postgres and sqlite to dynamically require their dependencies in the initialize function.